### PR TITLE
Segmentation docs and tests 1888

### LINKED ISSUE #1888: Semantic segmentation - Tests & Docs
Continues #1039 

Tasks:
- Add tests for SegmentationData - 0.5
- Add tests for checks with elif for segmentation (with working model) - 0.5
- Add tests for properties - 0.5
- Update checks docstrings - 0.5
- Add use-case of semantic segmentation to docs - 3 SP

Moved some tasks to #1939

### PR DESCRIPTION
**Pull request by nirhutnik**
Added segmentation data tutorial
(will also have tests)

closes #1888

### CODE REVIEW

**Comment by nirhutnik** [comment]
PDF of docs:
[Semantic Segmentation Tutorial — gallery-boilerplate documentation.pdf](https://github.com/deepchecks/deepchecks/files/9540849/Semantic.Segmentation.Tutorial.gallery-boilerplate.documentation.pdf)


Image that belongs there (missing in docs because of gallery-boilerplate bug):
segmentation_5_examples

**Review by noamzbr — COMMENTED** [review]
Cool! Added a few comments

Need to also mention segmentation in https://docs.deepchecks.com/stable/getting-started/welcome.html#computer-vision (Are there other places where we should mention this new task?)

File: deepchecks/vision/metrics_utils/semantic_segmentation_metrics.py [hunk 1]
```diff
@@ -51,10 +51,12 @@ def update(self, output: Tuple[torch.Tensor, torch.Tensor]):
     def compute(self):
         """Compute metric value."""
         sorted_classes = [int(class_id) for class_id in sorted(self._evals.keys())]
-        ret = []
+        max_class = max(sorted_classes)
+        scores_per_class = np.empty(max_class + 1)*np.nan
```
[the remark this answers is not in the document]
Comment (nirhutnik) [reply deepchecks/vision/metrics_utils/semantic_segmentation_metrics.py:55]: That all classes (that we know of) are there, yes.
This is done in other metrics as well

File: docs/source/user-guide/vision/quickstarts/plot_segmentation_tutorial.py [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+===============================
+
+In this tutorial, you will learn how to validate your **semantic segmentation model** using deepchecks test suites.
+You can read more about the different checks and suites for computer vision use cases at the
+:doc:`examples section  </checks_gallery/vision/index>`
+
+If you just want to see the output of this tutorial, jump to :ref:`observing_the_result` section.
+
+A semantic segmentation task is a task where every pixel of the image is labeled with a single class.
+Therefore, the common output of these tasks is an image of identical size to the input, with a vector for each pixel
```
[the remark this answers is not in the document]
Comment (nirhutnik) [reply docs/source/user-guide/vision/quickstarts/plot_segmentation_tutorial.py:15]: Hmmm. I'm not sure about that. And not really sure how important that is 😅 
Can change to "a common output" if you want

File: docs/source/user-guide/vision/quickstarts/plot_segmentation_tutorial.py [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+import matplotlib.pyplot as plt
+import numpy as np
+import torch
+import torchvision.transforms.functional as F
+
+
+# %%
+# Load Data
+# ~~~~~~~~~
+# The model in this tutorial is used to detect different objects in images, based on the Pascal VOC dataset.
```
[the remark this answers is not in the document]
Comment (nirhutnik) [reply docs/source/user-guide/vision/quickstarts/plot_segmentation_tutorial.py:41]: This refers to the model. The model is trained on Pascal VOC.

File: docs/source/user-guide/vision/quickstarts/plot_segmentation_tutorial.py [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+for i, img in enumerate(masked_images):
+    img = img.detach()
+    img = F.to_pil_image(img)
+    axs[i].imshow(np.asarray(img))
+    axs[i].set(xticklabels=[], yticklabels=[], xticks=[], yticks=[])
+
+fix.show()
+
+#%%
+# .. image :: /_static/images/tutorials/segmentation_5_examples.png
```
[the remark this answers is not in the document]
Comment (nirhutnik) [reply docs/source/user-guide/vision/quickstarts/plot_segmentation_tutorial.py:78]: It's the same, but for some reason that I didn't try to understand it doesn't show. This is like this in the detection tutorial as well so I just followed the same logic.

File: docs/source/user-guide/vision/quickstarts/plot_segmentation_tutorial.py [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+batch = next(iter(train_loader))
+
+print("Batch type is: ", type(batch))
+print("First element is: ", type(batch[0]), "with len of ", len(batch[0]))
+print("Example output of an image shape from the dataloader ", batch[0][0].shape)
+print("Image values", batch[0][0])
+print("-" * 80)
+
+print("Second element is: ", type(batch[1]), "with len of ", len(batch[1]))
+print("Example output of a label shape from the dataloader ", batch[1][0])
```
[the remark this answers is not in the document]
Comment (nirhutnik) [reply docs/source/user-guide/vision/quickstarts/plot_segmentation_tutorial.py:110]: Sure why not

File: docs/source/user-guide/vision/quickstarts/plot_segmentation_tutorial.py [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+             8: 'cat', 9: 'chair', 10: 'cow', 11: 'dining table', 12: 'dog', 13: 'horse', 14: 'motorcycle',
+             15: 'person', 16: 'potted plant', 17: 'sheep', 18: 'couch', 19: 'train', 20: 'tv'}
+
+training_data = CocoSegmentationData(data_loader=train_loader, label_map=LABEL_MAP)
+test_data = CocoSegmentationData(data_loader=test_loader, label_map=LABEL_MAP)
+
+training_data.validate_format(model, device=device)
+test_data.validate_format(model, device=device)
+
+# And observe the output:
```
[the remark this answers is not in the document]
Comment (nirhutnik) [reply docs/source/user-guide/vision/quickstarts/plot_segmentation_tutorial.py:174]: Ummm not it's not?
image
Or did I misunderstand you?

File: docs/source/user-guide/vision/quickstarts/plot_segmentation_tutorial.py [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+
+# %%
+# Running Deepchecks' full suite on our data and model!
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+# Now that we have defined the task class, we can validate the model with the full suite of deepchecks.
+# This can be done with this simple few lines of code:
+
+from deepchecks.vision.suites import full_suite
```
[the remark this answers is not in the document]
Comment (nirhutnik) [reply docs/source/user-guide/vision/quickstarts/plot_segmentation_tutorial.py:182]: Honestly not sure. It won't be very interesting as this runs on COCO which is a very limited dataset.

File: docs/source/user-guide/vision/quickstarts/plot_segmentation_tutorial.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (noamzbr) [reply docs/source/user-guide/vision/quickstarts/plot_segmentation_tutorial.py:41]: So from here it reads as if the images are based on Pascal VOC dataset, not the model.

File: docs/source/user-guide/vision/quickstarts/plot_segmentation_tutorial.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (noamzbr) [reply docs/source/user-guide/vision/quickstarts/plot_segmentation_tutorial.py:110]: I mean, did you intend to show the shape or the data itself?
(I'm for the shape because the data is big)

File: docs/source/user-guide/vision/quickstarts/plot_segmentation_tutorial.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (noamzbr) [reply docs/source/user-guide/vision/quickstarts/plot_segmentation_tutorial.py:182]: We don't want to showcase the full suite as we don't want people to use the full suite, and rather use the individual suite according to their stage in the model dev cycle.

File: docs/source/user-guide/vision/quickstarts/plot_segmentation_tutorial.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (noamzbr) [reply docs/source/user-guide/vision/quickstarts/plot_segmentation_tutorial.py:174]: Ahh because it's one code cell
my bad

File: docs/source/user-guide/vision/quickstarts/plot_segmentation_tutorial.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (noamzbr) [reply docs/source/user-guide/vision/quickstarts/plot_segmentation_tutorial.py:15]: Maybe, not critical but caught my eye

**Review by noamzbr — CHANGES_REQUESTED** [review]
Important comment! Need to add SegmentationData to `deepchecks.vision.rst` index file

File: docs/source/user-guide/vision/quickstarts/plot_segmentation_tutorial.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (nirhutnik) [reply docs/source/user-guide/vision/quickstarts/plot_segmentation_tutorial.py:41]: Disagree but fixed

File: docs/source/user-guide/vision/quickstarts/plot_segmentation_tutorial.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (nirhutnik) [reply docs/source/user-guide/vision/quickstarts/plot_segmentation_tutorial.py:182]: Changed to model_evaluation

File: docs/source/user-guide/vision/quickstarts/plot_segmentation_tutorial.py [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+
+#%%
+# .. image :: /_static/images/tutorials/segmentation_5_examples.png
+#     :alt: COCO images with segmenation labels
+#
+# Downloading a Pre-trained Model
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+# In this tutorial, we will download a pre-trained LRSAPP model and a MobileNetV3 Large backbone
+# from the official PyTorch repository. For more details, please refer to the
+# `official documentation <https://pytorch.org/vision/main/models/generated/torchvision.models.segmentation.lraspp_mobilenet_v3_large.html>`_.
```
Comment (shir22) [inline docs/source/user-guide/vision/quickstarts/plot_segmentation_tutorial.py:85]:
```suggestion
# `official documentation <https://pytorch.org/vision/main/models/generated/torchvision.models.segmentation.lraspp_mobilenet_v3_large.html>`__.
```

File: docs/source/user-guide/vision/quickstarts/plot_segmentation_tutorial.py [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+test_data = CocoSegmentationData(data_loader=test_loader, label_map=LABEL_MAP)
+
+training_data.validate_format(model, device=device)
+test_data.validate_format(model, device=device)
+
+# And observe the output:
+
+# %%
+# Running Deepchecks' model evaluation suite on our data and model!
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Comment (shir22) [inline docs/source/user-guide/vision/quickstarts/plot_segmentation_tutorial.py:178]:
```suggestion
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

File: docs/source/user-guide/vision/quickstarts/plot_segmentation_tutorial.py [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+    img = img.detach()
+    img = F.to_pil_image(img)
+    axs[i].imshow(np.asarray(img))
+    axs[i].set(xticklabels=[], yticklabels=[], xticks=[], yticks=[])
+
+fix.show()
+
+#%%
+# .. image :: /_static/images/tutorials/segmentation_5_examples.png
+#     :alt: COCO images with segmenation labels
```
Comment (shir22) [inline docs/source/user-guide/vision/quickstarts/plot_segmentation_tutorial.py:79]:
```suggestion
#     :alt: COCO images with segmentation labels
```

File: docs/source/user-guide/vision/quickstarts/plot_segmentation_tutorial.py [hunk 11]
```diff
@@ -0,0 +1,199 @@
+"""
+.. _vision_segmentation_tutorial:
+
+===============================
+Semantic Segmentation Tutorial
+===============================
+
+In this tutorial, you will learn how to validate your **semantic segmentation model** using deepchecks test suites.
+You can read more about the different checks and suites for computer vision use cases at the
+:doc:`examples section  </checks_gallery/vision/index>`
```
Comment (shir22) [inline docs/source/user-guide/vision/quickstarts/plot_segmentation_tutorial.py:10]:
```suggestion
:doc:`examples section  </checks_gallery/vision/index>`.
```

File: docs/source/user-guide/vision/quickstarts/plot_segmentation_tutorial.py [hunk 12]
```diff
@@ -0,0 +1,199 @@
...
+# Now that we have defined the task class, we can validate the model with the model evaluation suite of deepchecks.
+# This can be done with this simple few lines of code:
+
+from deepchecks.vision.suites import model_evaluation
+
+suite = model_evaluation()
+result = suite.run(training_data, test_data, model, device=device)
+
+# %%
+# .. _observing_the_result:
```
Comment (shir22) [inline docs/source/user-guide/vision/quickstarts/plot_segmentation_tutorial.py:188]:
```suggestion
# .. _vision_segmentation_tutorial__observing_the_result:
```

File: docs/source/user-guide/vision/quickstarts/plot_detection_tutorial.py [hunk 13]
```diff
@@ -11,7 +11,7 @@
 
 If you just want to see the output of this tutorial, jump to :ref:`observing_the_result` section.
```
Comment (shir22) [inline docs/source/user-guide/vision/quickstarts/plot_detection_tutorial.py:12]:
```suggestion
If you just want to see the output of this tutorial, jump to :ref:`vision_segmentation_tutorial__observing_the_result` section.
```
(added comment also in relevant place)

alternatively, you can do `#observing-the-results`__ or 'observing the results <#observing-the-results>`__
without defining the label (and delete the definition)

anyway labels are project-wide so if you do choose to leave it that way, we add a file-name-related prefix for uniqueness...

**Review by noamzbr — APPROVED** [review]
Approved pending on @shir22 s comments
